### PR TITLE
move BigQuery backend

### PR DESCRIPTION
**Pull request by vijoin**
Related to issue #2417

### CODE REVIEW

File: ibis/backends/bigquery/tests/test_client.py [hunk 1]
```diff
@@ -18,8 +18,8 @@
 ga = pytest.importorskip('google.auth')
 exceptions = pytest.importorskip('google.api_core.exceptions')
 
-from ibis.bigquery.client import bigquery_param  # noqa: E402, isort:skip
-from ibis.bigquery.tests.conftest import connect  # noqa: E402, isort:skip
+from ibis.backends.bigquery.client import bigquery_param  # noqa: E402, isort:skip
+from ibis.backends.bigquery.tests.conftest import connect  # noqa: E402, isort:skip
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_client.py:22]:
```suggestion
from .client import bigquery_param  # noqa: E402, isort:skip
from .tests.conftest import connect  # noqa: E402, isort:skip
```

File: ibis/backends/bigquery/tests/test_client.py [hunk 2]
```diff
@@ -75,7 +75,7 @@ def test_compile_toplevel():
 
     # it works!
     expr = t.foo.sum()
-    result = ibis.bigquery.compile(expr)
+    result = ibis.backends.bigquery.compile(expr)
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_client.py:78]:
```suggestion
    result = ibis.bigquery.compile(expr)
```
Quite misleading, I know, but this is the `from ibis.backends import bigquery` in `ibis/__init__.py`, not the module itself.

File: ibis/backends/bigquery/tests/test_compiler.py [hunk 3]
```diff
@@ -4,7 +4,7 @@
 import pytest
 
 import ibis
-import ibis.bigquery as bq
+import ibis.backends.bigquery as bq
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_compiler.py:7]:
```suggestion
import . as bq
```
I guess this works

File: ibis/backends/bigquery/tests/test_compiler.py [hunk 4]
```diff
@@ -102,7 +102,7 @@ def test_to_timestamp(alltypes, timezone, project_id):
 )
 def test_literal_date(case, expected, dtype):
     expr = ibis.literal(case, type=dtype).year()
-    result = ibis.bigquery.compile(expr)
+    result = ibis.backends.bigquery.compile(expr)
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_compiler.py:105]:
```suggestion
    result = ibis.bigquery.compile(expr)
```

File: ibis/backends/bigquery/tests/test_compiler.py [hunk 5]
```diff
@@ -145,14 +145,14 @@ def test_literal_date(case, expected, dtype):
 def test_day_of_week(case, expected, dtype, strftime_func):
     date_var = ibis.literal(case, type=dtype)
     expr_index = date_var.day_of_week.index()
-    result = ibis.bigquery.compile(expr_index)
+    result = ibis.backends.bigquery.compile(expr_index)
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_compiler.py:148]:
```suggestion
    result = ibis.bigquery.compile(expr_index)
```
All the ones below too.

File: ibis/backends/bigquery/tests/test_datatypes.py [hunk 6]
```diff
@@ -3,7 +3,7 @@
 from pytest import param
 
 import ibis.expr.datatypes as dt
-from ibis.bigquery.datatypes import (
+from ibis.backends.bigquery.datatypes import (
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_datatypes.py:6]:
```suggestion
from .datatypes import (
```

File: ibis/backends/bigquery/tests/udf/test_core.py [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 import pytest
 
-from ibis.bigquery.udf.core import PythonToJavaScriptTranslator, SymbolTable
+from ibis.backends.bigquery.udf.core import PythonToJavaScriptTranslator, SymbolTable
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/udf/test_core.py:7]:
```suggestion
from .udf.core import PythonToJavaScriptTranslator, SymbolTable
```

File: ibis/backends/bigquery/tests/udf/test_find.py [hunk 8]
```diff
@@ -2,7 +2,7 @@
 
 import pytest
 
-from ibis.bigquery.udf.find import find_names
+from ibis.backends.bigquery.udf.find import find_names
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/udf/test_find.py:5]:
```suggestion
from .udf.find import find_names
```

File: ibis/backends/bigquery/tests/udf/test_udf_execute.py [hunk 9]
```diff
@@ -7,7 +7,7 @@
 
 import ibis
 import ibis.expr.datatypes as dt
-from ibis.bigquery import udf  # noqa: E402
+from ibis.backends.bigquery import udf  # noqa: E402
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/udf/test_udf_execute.py:10]:
```suggestion
from . import udf  # noqa: E402
```

File: ibis/backends/bigquery/tests/udf/test_udf_execute.py [hunk 10]
```diff
@@ -26,7 +26,7 @@
 
 @pytest.fixture(scope='module')
 def client():
-    from ibis.bigquery.tests.conftest import connect
+    from ibis.backends.bigquery.tests.conftest import connect
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/udf/test_udf_execute.py:29]:
```suggestion
    from .tests.conftest import connect
```

File: ibis/backends/bigquery/udf/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+from .api import udf  # noqa: F401
```
Comment (datapythonista) [inline ibis/backends/bigquery/udf/__init__.py:1]: Can you just rename `api.py` to `__init__.py` instead?

**Review by datapythonista — COMMENTED** [review]
Thanks @vijoin, there are few things quite misleading in Ibis, and there are some imports in your PR that are not correct, but nice PR, other than that looks good. Thanks for the work.

**Comment by datapythonista** [comment]
Also, if you can have a look at the error about the lines too long, that would be great.

**Comment by vijoin** [comment]
Those long line errors are produced by inline comments, and I think those
are for flake8 purpose.

File: docs/source/backends/bigquery.rst [hunk 12]
```diff
@@ -53,7 +53,7 @@ method of :class:`ibis.bigquery.client.BigQueryClient` objects:
 
 API
 ---
-.. currentmodule:: ibis.bigquery.api
+.. currentmodule:: ibis.backends.bigquery.api
```
Comment (datapythonista) [inline docs/source/backends/bigquery.rst:56]:
```suggestion
.. currentmodule:: ibis.backends.bigquery
```

File: ibis/backends/bigquery/tests/conftest.py [hunk 13]
```diff
@@ -39,7 +39,7 @@ def connect(project_id, dataset_id, application_name=None):
         google_application_credentials
     )
     try:
-        return ibis.bigquery.connect(
+        return ibis.backends.bigquery.connect(
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/conftest.py:42]:
```suggestion
        return ibis.bigquery.connect(
```

File: ibis/backends/bigquery/tests/test_compiler.py [hunk 14]
```diff
@@ -152,7 +152,7 @@ def test_day_of_week(case, expected, dtype, strftime_func):
     )
 
     expr_name = date_var.day_of_week.full_name()
-    result = ibis.bigquery.compile(expr_name)
+    result = ibis.backends.bigquery.compile(expr_name)
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_compiler.py:155]:
```suggestion
    result = ibis.bigquery.compile(expr_name)
```
In this one and all the rest, this is not the `ibis.bigquery`/`ibis.backends.bigquery` module. Instead, it's the attribute `bigquery` of ibis (in `ibis/__init__.py`). Which is that module actually, but created as an attribute of the module ibis. Very confusing, I know, but luckily, with this PR there will be no ambiguity anymore (that's one of the reasons for the move of the backend).

**Review by datapythonista — COMMENTED** [review]
Thanks @vijoin, great changes. There are just couple of places where things are confusing and the imports are not correct, but other than that looks good.

File: ibis/backends/bigquery/tests/test_compiler.py (hunk 14 above)
Comment (vijoin) [reply ibis/backends/bigquery/tests/test_compiler.py:155]: Great! I get it now!

File: ibis/backends/bigquery/tests/test_compiler.py [hunk 15]
```diff
@@ -186,7 +186,7 @@ def test_day_of_week(case, expected, dtype, strftime_func):
 )
 def test_literal_timestamp_or_time(case, expected, dtype):
     expr = ibis.literal(case, type=dtype).hour()
-    result = ibis.bigquery.compile(expr)
+    result = ibis.backends.bigquery.compile(expr)
```
Comment (datapythonista) [inline ibis/backends/bigquery/tests/test_compiler.py:189]: This and the other 7 similar changes in this file also need to be reverted.

Comment (vijoin) [reply ibis/backends/bigquery/tests/test_compiler.py]: done!

File: ibis/backends/bigquery/udf/__init__.py (hunk 11 above)
Comment (vijoin) [reply ibis/backends/bigquery/udf/__init__.py]: done!

**Review by datapythonista — APPROVED** [review]
Thanks @vijoin, looks great. We need to fix the problem in the CI and get a green build before we can merge, but this should be ready.

**Comment by datapythonista** [comment]
@vijoin CI problems should be fixed now. You'll have to merge master into your branch, fix the conflict, and also change the module in https://github.com/ibis-project/ibis/blob/master/ci/recipe/meta.yaml#L64 to get the CI passing (ibis.bigquery -> ibis.backends.bigquery).

After that we can merge this. Thanks a lot for the work on this PR!

**Comment by datapythonista** [comment]
@vijoin do you have time to merge master and update `ci/recipe/meta.yaml`? You've also got a commit that needs to be fixed now.

**Comment by vijoin** [comment]
Sure. I'll do it in a few minutes

**Comment by jreback** [comment]
thanks @vijoin

**Pull request merged**